src/runtime_src/driver/xclng/drm/xocl/subdev/mb_scheduler.c

```diff
@@ -622,69 +628,131 @@ cleanup_exec(struct xocl_cmd *xcmd)
+ */
+static void
+exec_cfg(struct exec_core *exec)
+{
+}
+
+/**
+ * stop_exec() - Stop the scheduler from scheduling commands on this core
  *
- * Clear stale command objects associated with execution core.
- * This can occur if the HW for some reason hangs.
+ * @exec:  Execution core (device) to stop
+ *
+ * Block access to current exec_core (device).  This API must be called prior
+ * to performing an AXI reset and downloading of a new xclbin.  Calling this
+ * API flushes the commands running on current device and prevents new
+ * commands from being scheduled on the device.  This effectively prevents any
+ * further commands from running on the device
  */
 SCHED_UNUSED
 static void
-reset_exec(struct exec_core* exec)
+exec_stop(struct exec_core* exec)
 {
-	int i;
+	int idx;
 	struct list_head *pos, *next;
+	struct xocl_dev *xdev = exec_get_xdev(exec);
 
-	userpf_info(exec_get_xdev(exec),"reset_exec(%p)\n",exec);
+	mutex_lock(&exec->exec_lock);
 
-	/* clear stale command objects if any */
+	userpf_info(xdev,"exec_stop(%p)\n",exec);
+
+	// flush command objects if any (hanging)
 	list_for_each_safe(pos, next, &pending_cmds) {
 		struct xocl_cmd *xcmd = list_entry(pos,struct xocl_cmd,list);
 		if (xcmd->exec!=exec)
 			continue;
-		DRM_INFO("deleting stale pending cmd\n");
+		userpf_info(xdev,"deleting stale pending cmd\n");
 		cleanup_exec(xcmd);
 	}
 	list_for_each_safe(pos, next, &global_scheduler0.command_queue) {
 		struct xocl_cmd *xcmd = list_entry(pos,struct xocl_cmd,list);
 		if (xcmd->exec!=exec)
 			continue;
-		DRM_INFO("deleting stale scheduler cmd\n");
+		userpf_info(xdev,"deleting stale scheduler cmd\n");
 		cleanup_exec(xcmd);
 	}
 
-	/* reset exec state */
-        for (i=0; i<MAX_SLOTS; ++i)
-		exec->submitted_cmds[i] = NULL;
+	// Stale commands were flushed, reset submitted command state
+        for (idx=0; idx<MAX_SLOTS; ++idx)
+		exec->submitted_cmds[idx] = NULL;
 
-	exec->num_slots = 16;
+	memset(exec->slot_status,0,MAX_U32_SLOT_MASKS*sizeof(u32));
+	exec->slot_status[0] = 1; // reserve for ctrl commands
+	exec->ctrl_busy=false;
+	exec->stopped = true;
+
+	mutex_unlock(&exec->exec_lock);
+}
+
+/**
+ * exec_reset() - Reset the scheduler
+ *
+ * @exec: Execution core (device) to reset
+ *
+ * TODO: Perform scheduler configuration based on current xclbin
+ *       rather than relying of cfg command
+ */
+SCHED_UNUSED
+static void
+exec_reset(struct exec_core* exec)
+{
+	struct xocl_dev *xdev = exec_get_xdev(exec);
+
+	mutex_lock(&exec->exec_lock);
+
+	userpf_info(xdev,"exec_reset(%p)\n",exec);
+
+	// only reconfigure the scheduler on new xclbin
+	if (uuid_equal(&exec->xclbin_id, &xdev->xclbin_id) && exec->configured) {
+		exec->stopped = false;
+		exec->configured = false;  // TODO: remove, but hangs ERT because of in between AXI resets
+		goto out;
+	}
+
+	userpf_info(xdev,"exec_reset resets for new xclbin");
+	memset(exec->cu_usage,0,MAX_CUS*sizeof(u32));
+	uuid_copy(&exec->xclbin_id, &xdev->xclbin_id);
```

Comment:
What cache?   There is no complexity comparing xclbin ids.   The scheduler cannot be configured without an xclbin, it needs meta data from the xclbin.   It would remain unconfigured in your use case.